Add usage quota to AI creation of flashcards and collections

sebampuero opened the pull request:
This PR adds usage quotas to AI creation of flashcards and collections.

### Backend

- Usage is user based. Each user has a usage quota.
- Usage is configurable: time frame (in days) and count. 
- Adds tests to backend functionality.

### Frontend

- Shows percentage usage and next refresh date in the AI creation dialog (ala Copilot)
- Does not allow AI creation when quota is equal to 100%

0010aor commented:
Thanks a lot for the contribution it looks good at first glance. I'll take a closer look over the weekend. Appreciate it!

0010aor commented:
## Suggested Improvements

### 1. **Unique Constraint on Quota**
- **Suggestion:** Add a unique constraint on `user_id` in the `AIUsageQuota` model to ensure only one quota record per user. This will prevent data integrity issues and simplify logic.

### 2. **Configuration Validation**
- **Suggestion:** Validate that `AI_MAX_USAGE_QUOTA` and `AI_QUOTA_TIME_RANGE_DAYS` are set and positive at startup. This will prevent misconfiguration and runtime errors (division by zero in the percentage calculation).

### 3. **Single function**
- **Suggestion:** Consolidate quota logic into a single, clear function that handles creation, reset, increment, and enforcement.

### 4. **No Unnecessary Commits**
- **Suggestion:** Only commit the session when changes are actually made to the quota record.

### 5. **Atomic Quota Increment**
- **Suggestion:** Use a database-level atomic update for incrementing the quota. This will prevent race conditions and ensure that two concurrent requests can't both increment the quota past the allowed maximum.

### 6. **Move Endpoint to Users Folder**
- **Suggestion:** Move the AI usage quota endpoint to the users folder (e.g., `/users/me/ai-usage-quota`) This keeps user-related logic together.

BTW, if you are ok with the new changes to the model please revert the latest migration and autogenerate a new one.

to apply the suggestions locally (please keep it mind I have not tested the suggestions) run the following:
```
curl -L https://gist.githubusercontent.com/0010aor/98d8d8c7f54c3e36341d9b685a274291/raw -o change.diff && git apply change.diff
```

thanks a lot for the contribution! 👐

0010aor commented:
![image](https://github.com/user-attachments/assets/a2a99d9c-a245-49e3-8619-de7588939bbf)

## Frontend Dialog Suggestions

- **Simplify the Quota Display:** Instead of showing both the reset date and percentage used (e.g., "AI Usage quota reset date: 09/06/2025" and "AI Usage quota used: 6%"), consider setting `AI_QUOTA_TIME_RANGE_DAYS` to 1 so the quota resets daily. Then, in the dialog, you can show a single line such as:
  - `Daily usage: X requests left`
- **Possible Bug:** If a user used AI yesterday, and today is a new day, the dialog might still show usage from yesterday (e.g., "6% used") when it should be reset to 0%. Make sure the UI reflects the reset correctly at the start of each new quota window.
- **Show Requests Left:** Displaying the number of requests left for today (e.g., `Daily usage: 24 requests left`) is likely clearer and more actionable for users than a percentage.

sebampuero commented:
@0010aor thanks for the feedback! I will take a look these days

sebampuero commented:
Hi @0010aor 
I implemented some of the suggested changes:
### 1. Unique Constraint on Quota
_Suggestion: Add a unique constraint on user_id in the AIUsageQuota model to ensure only one quota record per user. This will prevent data integrity issues and simplify logic_. ✅
### 2. Configuration Validation
_Suggestion: Validate that AI_MAX_USAGE_QUOTA and AI_QUOTA_TIME_RANGE_DAYS are set and positive at startup. This will prevent misconfiguration and runtime errors (division by zero in the percentage calculation)._ ✅
### 3. Single function
_Suggestion: Consolidate quota logic into a single, clear function that handles creation, reset, increment, and enforcement._ ✅
I modified the suggested function slightly. It had a bug because it wasn't updating the quota. The optimistic locking is maintained.
### 4. No Unnecessary Commits
_Suggestion: Only commit the session when changes are actually made to the quota record_. ✅
### 5. Atomic Quota Increment
_Suggestion: Use a database-level atomic update for incrementing the quota. This will prevent race conditions and ensure that two concurrent requests can't both increment the quota past the allowed maximum._ ✅ optimistic lock is maintained as explained above so that concurrent requests don't modify the quota usage in a weird manner.
### 6. Move Endpoint to Users Folder
_Suggestion: Move the AI usage quota endpoint to the users folder (e.g., /users/me/ai-usage-quota) This keeps user-related logic together._ ✅

## Frontend Dialog Suggestions
### Simplify the Quota Display: 
_Instead of showing both the reset date and percentage used (e.g., "AI Usage quota reset date: 09/06/2025" and "AI Usage quota used: 6%"), consider setting AI_QUOTA_TIME_RANGE_DAYS to 1 so the quota resets daily. Then, in the dialog, you can show a single line such as:
Daily usage: X requests left_ ⚠️

### Possible Bug: 
_If a user used AI yesterday, and today is a new day, the dialog might still show usage from yesterday (e.g., "6% used") when it should be reset to 0%. Make sure the UI reflects the reset correctly at the start of each new quota window._ ✅ 
### Show Requests Left: 
_Displaying the number of requests left for today (e.g., Daily usage: 24 requests left) is likely clearer and more actionable for users than a percentage._ ✅ 

Note: I did not implement the first suggestion entirely. I did remove the usage percentage indeed but left the reset date intact. I based it on how copilow shows usage in VSCode:
![image](https://github.com/user-attachments/assets/d08d38dc-3ded-44aa-ba93-3fde3716e5b3)

This way the configuration can be dynamic, in essence, we can change to monthly, weekly, bi-weekly, etc.. usage by modifying the `AI_QUOTA_TIME_RANGE_DAYS` env var. If you think daily usage looks and feels better, I can set it to 1 so that daily usage is shown in the UI, that wouldn't be a problem 😊

0010aor commented on backend/src/core/config.py:
```diff
@@ -64,6 +64,8 @@ def SQLALCHEMY_DATABASE_URI(self) -> PostgresDsn:
 
     AI_API_KEY: str | None = None
     AI_MODEL: str | None = None
+    AI_MAX_USAGE_QUOTA: int | None = None
```
```
    AI_MAX_USAGE_QUOTA: int = 30
    AI_QUOTA_TIME_RANGE_DAYS: int = 1
```
so the project can run without requiring to define those variables in the .env

0010aor commented on backend/src/flashcards/api.py:
```diff
@@ -52,6 +53,10 @@ async def create_collection(
 
     if collection_in.prompt:
         try:
+            if not check_and_increment_ai_usage_quota(session, current_user):
```
i think we should use if not await asyncio.to_thread here right?

0010aor commented on backend/src/flashcards/api.py:
```diff
@@ -145,6 +150,10 @@ async def create_card(
     if not access_checked:
         raise HTTPException(status_code=404, detail="Collection not found")
     if card_in.prompt:
+        if not check_and_increment_ai_usage_quota(session, current_user):
```
same here?

0010aor commented on frontend/src/components/commonUI/AiPromptDialog.tsx:
```diff
@@ -37,6 +40,16 @@ const AiPromptDialog: React.FC<AiDialogProps> = ({
+  const usageQuota: AIUsageQuota = data || {
+    max_usage_allowed: 0,
+    usage_count: 0,
+    reset_date: new Date().toDateString(),
```
The reset date is displayed using toLocaleDateString(). Ensure the date is always in a valid format and consider time zone differences.

```
reset_date: new Date().toISOString(),

```

0010aor reviewed (commented):
added few small comments, rest looks good to me

sebampuero commented:
@0010aor I added the suggested changes. I used `reset_date: new Date().toLocaleDateString(),``because `toLocaleDateString()` uses the default browser's locale, respecting the user's locale
